File: static/scripts/path_find/DFS.js
import { statusCell } from "../StatusCell.js"
import {renderPath, resetCellColors, animateVisitedCells, isRunning, setRunningState} from "./SharedFunction.js"
import { updateInfor, scrollToBottom, showToast } from "./InformationPanel.js";

document.getElementById('DFSButton').addEventListener('click', async () => {
    if (isRunning) {
            showToast('An algorithm is already running. Please wait!');
            return; // Prevent starting a new algorithm
        }
    
    resetCellColors();
    showToast('Depth-first Search is unweighted and does not guarantee the shortest path!');
    const maze = statusCell()


    const data = {
        maze: maze,
        start: [startCell.row, startCell.col],
        end: [endCell.row, endCell.col],

    };

    try {
        setRunningState(true);
        const response = await axios.post('/pathfinding/dfs', data);
        const { path, visited_cells, time } = response.data;
        updateInfor("DFS", visited_cells, time, path)
        scrollToBottom();
        await animateVisitedCells(visited_cells);

        // After visited cells are animated, render the path
        renderPath(path);
    } catch (error) {
        console.error('Error generating maze:', error);
    } finally {
        setRunningState(false); // Reset the flag after completion
    }
});

File: pathfinding/dfs.py
class Depth_First_Search:
    def __init__(self, maze):
        self.maze = maze

    def search(self, start, end):
        self.visited = set()
        self.path = []
        self.visited_cells = []
        self.dfs(start, start, end)  
        if not self.path:
            return None
        return self.path, self.visited_cells

    def dfs(self, current, start, end):
        if current.x == end.x and current.y == end.y:
            self.path.append((current.x, current.y))  
            return True

        self.visited.add((current.x, current.y))
        self.visited_cells.append((current.x, current.y))

        for direction in ['up', 'right', 'down', 'left']:
            next_cell = getattr(current, direction)()  
            if ((next_cell.x, next_cell.y) not in self.visited and self.maze.is_valid(next_cell, start, end)):
                if self.dfs(next_cell, start, end):
                    self.path.append((current.x, current.y))  
                    return True

        return False


    
    